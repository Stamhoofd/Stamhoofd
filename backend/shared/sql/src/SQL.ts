import { SQLExpression } from "./SQLExpression";
import { SQLSelect } from "./SQLSelect";
import { SQLColumnExpression, SQLTableExpression, SQLWildcardSelectExpression } from "./SQLExpressions";
import { SQLJoin, SQLJoinType } from "./SQLJoin";

class StaticSQL {
    wildcard(namespace?: string) {
        return new SQLWildcardSelectExpression(namespace)
    }

    column(namespace: string, column: string);
    column(column: string);
    column(namespaceOrColumn: string, column?: string) {
        if (column === undefined) {
            return new SQLColumnExpression(namespaceOrColumn)
        }
        return new SQLColumnExpression(namespaceOrColumn, column)
    }

    table(namespace: string, table: string);
    table(table: string);
    table(namespaceOrTable: string, table?: string) {
        if (table === undefined) {
            return new SQLTableExpression(namespaceOrTable)
        }
        return new SQLTableExpression(namespaceOrTable, table)
    }

    select(...columns: SQLExpression[]) {
        if (columns.length === 0) {
            return new SQLSelect(this.wildcard())
        }
        return new SQLSelect(...columns)
    }

    leftJoin(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Left, table)
    }

    rightJoin(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Right, table)
    }

    innerJoin(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Inner, table)
    }

    join(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Inner, table)
    }
}

export const SQL = new StaticSQL();