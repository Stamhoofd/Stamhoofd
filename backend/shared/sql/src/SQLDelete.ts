import { Database } from "@simonbackx/simple-database";
import { SQLExpression, SQLExpressionOptions, SQLQuery, joinSQLQuery, normalizeSQLQuery } from "./SQLExpression";
import { SQLJoin } from './SQLJoin';
import { SQLWhere, addWhereHelpers } from "./SQLWhere";

class DeleteBase implements SQLExpression {
    _from: SQLExpression;
    _where: SQLWhere|null = null;
    _joins: (InstanceType<typeof SQLJoin>)[] = [];

    constructor() {
    }

    clone(): this {
        const c = new SQLDelete()
        Object.assign(c, this);
        return c as any;
    }

    from(table: SQLExpression): this {
        this._from = table;
        return this;
    }

    join(join: InstanceType<typeof SQLJoin>): this {
        this._joins.push(join);
        return this;
    }

    getSQL(options?: SQLExpressionOptions): SQLQuery {
        const query: SQLQuery[] = [
            'DELETE'
        ]

        options = options ?? {}
        options.defaultNamespace = (this._from as any).namespace ?? (this._from as any).table ?? undefined;

        query.push(
            'FROM' 
        )

        query.push(this._from.getSQL(options));

        query.push(...this._joins.map(j => j.getSQL(options)))

        if (this._where) {
            query.push('WHERE')
            query.push(this._where.getSQL(options))
        }

        return joinSQLQuery(query, ' ');
    }

    async delete(): Promise<{affectedRows: number}> {
        const {query, params} = normalizeSQLQuery(this.getSQL())

        console.log(query, params);
        const [rows] = await Database.delete(query, params);
        return rows;
    }

    async then(onFulfilled: (value: {affectedRows: number}) => any, onRejected: (reason: any) => any): Promise<any> {
        return this.delete().then(onFulfilled, onRejected);
    }
}

export const SQLDelete = addWhereHelpers(DeleteBase)
