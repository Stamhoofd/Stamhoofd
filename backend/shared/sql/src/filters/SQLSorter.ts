import { PlainObject } from "@simonbackx/simple-encoding";
import { SortDefinition, SortList } from "@stamhoofd/structures";

import { SQLOrderBy, SQLOrderByDirection } from "../SQLOrderBy";

export type SQLSortDefinition<T, B extends PlainObject = PlainObject> = SortDefinition<T, B> & {
    toSQL(direction: SQLOrderByDirection): SQLOrderBy
};

export type SQLSortDefinitions<T = any> = Record<string, SQLSortDefinition<T>>

export function compileToSQLSorter(sortBy: SortList, definitions: SQLSortDefinitions): SQLOrderBy {
    const sorters: SQLOrderBy[] = [];
    
    for (const s of sortBy) {
        const d = definitions[s.key];
        if (!d) {
            throw new Error('Unknown sort key ' + s.key)
        }

        sorters.push(d.toSQL(s.order));
    }

    if (sorters.length === 0) {
        throw new Error('No sortBy passed')
    }

    return SQLOrderBy.combine(sorters)
}
