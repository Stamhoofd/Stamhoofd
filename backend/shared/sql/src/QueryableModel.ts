import { Model, SQLResultNamespacedRow } from '@simonbackx/simple-database';
import { SQLSelect } from './SQLSelect';
import { SQL } from './SQL';
import { SQLDelete } from './SQLDelete';
import { SQLUpdate } from './SQLUpdate';
import { SQLInsert } from './SQLInsert';

export class QueryableModel extends Model {
    static select<T extends typeof Model>(this: T): SQLSelect<InstanceType<T>> {
        const transformer = (row: SQLResultNamespacedRow): InstanceType<T> => {
            const d = (this as T).fromRow(row[this.table] as any) as InstanceType<T> | undefined;

            if (!d) {
                throw new Error('EmailTemplate not found');
            }

            return d;
        };

        const select = new SQLSelect(transformer, SQL.wildcard());
        return select.from(SQL.table(this.table));
    }

    static delete(): SQLDelete {
        return SQL.delete().from(SQL.table(this.table));
    }

    static update(): SQLUpdate {
        return SQL.update(SQL.table(this.table));
    }

    static insert(): SQLInsert {
        return SQL.insert(SQL.table(this.table));
    }
}
