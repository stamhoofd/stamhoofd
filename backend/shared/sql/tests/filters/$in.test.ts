import { baseSQLFilterCompilers, createColumnFilter, SQLValueType } from '../../src/filters/SQLFilter';
import { SQL } from '../../src/SQL';
import { TableDefinition, test, testError, testMatch } from '../utils';

describe('$in', () => {
    /**
     * Tests that should be repeated for all filter types
     */
    describe('Common checks', () => {
        // todo
    });

    it('throws when not passing an array', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await testError({
            filter: {
                age: {
                    $in: 18,
                },
            },
            filters,
            error: 'Expected array at $in filter',
        });
    });

    it('throws when passing more than 1000 values to the filter', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await testError({
            filter: {
                age: {
                    $in: Array.from({ length: 1001 }, (_, i) => i + 1),
                },
            },
            filters,
            error: 'Too many values in $in filter, maximum is 1000',
        });
    });

    it('can filter on $in', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $in: [18, 19, 20],
                },
            },
            filters,
            query: {
                query: '`default`.`age` IN (?)',
                params: [
                    [18, 19, 20],
                ],
            },
        });
    });

    it('can filter on single $in', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $in: [18],
                },
            },
            filters,
            query: {
                query: '`default`.`age` = ?',
                params: [18],
            },
        });
    });

    it('can filter on empty $in', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $in: [],
                },
            },
            filters,
            query: {
                query: '',
                params: [],
            },
        });
    });

    it('can invert $in inside', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $not: {
                        $in: [18, 19, 20],
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` NOT IN (?)',
                params: [
                    [18, 19, 20],
                ],
            },
        });
    });

    it('can invert $in outside', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                $not: {
                    age: {
                        $in: [18, 19, 20],
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` NOT IN (?)',
                params: [
                    [18, 19, 20],
                ],
            },
        });
    });

    it('splits up when using null', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $in: [18, 19, 20, null],
                },
            },
            filters,
            query: {
                query: '`default`.`age` IS NULL OR `default`.`age` IN (?)',
                params: [
                    [18, 19, 20],
                ],
            },
        });
    });

    it('works when only including null', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $in: [null],
                },
            },
            filters,
            query: {
                query: '`default`.`age` IS NULL',
                params: [],
            },
        });
    });

    it('throws an error when trying to search in json objects', async () => {
        const filters = {
            ...baseSQLFilterCompilers,
            settings: createColumnFilter({ expression: SQL.column('settings'), type: SQLValueType.JSONObject, nullable: false }),
        };
        await testError({
            filter: {
                settings: {
                    $in: ['apple', 'banana'],
                },
            },
            filters,
            error: 'Cannot compare with a JSON object',
        });
    });

    describe('JSON', () => {
        const tableDefinition: TableDefinition = {
            settings: {
                type: 'json',
                nullable: true,
            },
        };
        const filters = {
            ...baseSQLFilterCompilers,
            'settings.randomValues': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.randomValues'), type: SQLValueType.JSONArray, nullable: true },
            ),
            'settings.name': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.name'), type: SQLValueType.JSONString, nullable: true },
            ),
        };

        it('Can check if a string is in a JSON string array, case insensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            randomValues: [
                                'apple',
                                'Banana',
                                'cherrY',
                            ],
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple', 'banana', 'cherry', 'other'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['apple', 'banana', 'other'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['Apple', 'banana', 'other'],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['orange', 'kiwi'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['pple'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['àpple'],
                        },
                    },
                ],
            });
        });

        it('Can check if a JSON string equals values, case insensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            name: 'aPple',
                        },
                    },
                    {
                        settings: {
                            name: 'apple',
                        },
                    },
                    {
                        settings: {
                            name: 'APPLE',
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.name': {
                            $in: ['apple', 'banana', 'other'],
                        },
                    },
                    {
                        'settings.name': {
                            $in: ['apple'],
                        },
                    },
                    {
                        'settings.name': {
                            $in: ['Apple', 'other'],
                        },
                    },
                ],
                doNotMatch: [
                ],
            });
        });

        it('Can check if an emoji is in a JSON string array', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            randomValues: [
                                '🍎',
                                '🍌',
                                '👩🏽‍🎤',
                            ],
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['🍎', '🍌', '👩🏽‍🎤', 'other'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['🍎', '🍌', 'other'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['👩🏽‍🎤', 'other'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['👩🏽‍🎤'],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['other'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['👩🏽'],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: ['👩🏻‍🎤'],
                        },
                    },
                ],
            });
        });

        it('Can check if null is in a JSON array', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            randomValues: [
                                null,
                                'Banana',
                                false,
                                0,
                            ],
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple', null],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple'],
                        },
                    },
                ],
            });
        });

        it('Including null will also check on key existence for json arrays', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: null,
                    },
                    {
                        settings: {},
                    },
                    {
                        settings: {
                            randomValues: null,
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple', null],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [null],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple'],
                        },
                    },
                ],
            });
        });

        it('If a json array exists and does not contain null, it is not included when checking for null', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            randomValues: [
                                'Banana',
                                false,
                                0,
                            ],
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: [false],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple', null],
                        },
                    },
                ],
            });
        });

        it('Including null also matches when a JSON path does not exist', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {},
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple', 'banana', 'cherry', null],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: ['apple', 'banana', 'cherry'],
                        },
                    },
                ],
            });
        });

        it('Can search numbers in a JSON array', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            randomValues: [
                                null,
                                'Banana',
                                false,
                                0,
                                5,
                                15,
                            ],
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: [5, 15],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [5],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [15],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [99, 15],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: [6, 10],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [99, 14],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [],
                        },
                    },
                ],
            });
        });

        it('Can search booleans in a JSON array', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            randomValues: [
                                null,
                                'Banana',
                                false,
                                1,
                            ],
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.randomValues': {
                            $in: [false, true],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [false],
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.randomValues': {
                            $in: [true],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [0],
                        },
                    },
                    {
                        'settings.randomValues': {
                            $in: [],
                        },
                    },
                ],
            });
        });
    });
});
