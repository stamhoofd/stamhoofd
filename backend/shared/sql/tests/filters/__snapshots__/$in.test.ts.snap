// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$in JSON Can check if a JSON string equals values, case insensitive: SQL Query for filter: {"settings.name":{"$in":["Apple","other"]}} 1`] = `
{
  "params": [
    [
      "apple",
      "other",
    ],
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) IN (?)",
}
`;

exports[`$in JSON Can check if a JSON string equals values, case insensitive: SQL Query for filter: {"settings.name":{"$in":["apple","banana","other"]}} 1`] = `
{
  "params": [
    [
      "apple",
      "banana",
      "other",
    ],
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) IN (?)",
}
`;

exports[`$in JSON Can check if a JSON string equals values, case insensitive: SQL Query for filter: {"settings.name":{"$in":["apple"]}} 1`] = `
{
  "params": [
    "apple",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4 ERROR ON ERROR) = ?",
}
`;

exports[`$in JSON Can check if a string is in a JSON string array, case insensitive: SQL Query for filter: {"settings.randomValues":{"$in":["Apple","banana","other"]}} 1`] = `
{
  "params": [
    "["apple","banana","other"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if a string is in a JSON string array, case insensitive: SQL Query for filter: {"settings.randomValues":{"$in":["apple","banana","cherry","other"]}} 1`] = `
{
  "params": [
    "["apple","banana","cherry","other"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if a string is in a JSON string array, case insensitive: SQL Query for filter: {"settings.randomValues":{"$in":["apple","banana","other"]}} 1`] = `
{
  "params": [
    "["apple","banana","other"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if a string is in a JSON string array, case insensitive: SQL Query for not match filter: {"settings.randomValues":{"$in":["orange","kiwi"]}} 1`] = `
{
  "params": [
    "["orange","kiwi"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if a string is in a JSON string array, case insensitive: SQL Query for not match filter: {"settings.randomValues":{"$in":["pple"]}} 1`] = `
{
  "params": [
    "["pple"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if a string is in a JSON string array, case insensitive: SQL Query for not match filter: {"settings.randomValues":{"$in":["√†pple"]}} 1`] = `
{
  "params": [
    "["√†pple"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if an emoji is in a JSON string array: SQL Query for filter: {"settings.randomValues":{"$in":["üçé","üçå","other"]}} 1`] = `
{
  "params": [
    "["üçé","üçå","other"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if an emoji is in a JSON string array: SQL Query for filter: {"settings.randomValues":{"$in":["üçé","üçå","üë©üèΩ‚Äçüé§","other"]}} 1`] = `
{
  "params": [
    "["üçé","üçå","üë©üèΩ‚Äçüé§","other"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if an emoji is in a JSON string array: SQL Query for filter: {"settings.randomValues":{"$in":["üë©üèΩ‚Äçüé§","other"]}} 1`] = `
{
  "params": [
    "["üë©üèΩ‚Äçüé§","other"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if an emoji is in a JSON string array: SQL Query for filter: {"settings.randomValues":{"$in":["üë©üèΩ‚Äçüé§"]}} 1`] = `
{
  "params": [
    "["üë©üèΩ‚Äçüé§"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if an emoji is in a JSON string array: SQL Query for not match filter: {"settings.randomValues":{"$in":["other"]}} 1`] = `
{
  "params": [
    "["other"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if an emoji is in a JSON string array: SQL Query for not match filter: {"settings.randomValues":{"$in":["üë©üèª‚Äçüé§"]}} 1`] = `
{
  "params": [
    "["üë©üèª‚Äçüé§"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if an emoji is in a JSON string array: SQL Query for not match filter: {"settings.randomValues":{"$in":["üë©üèΩ"]}} 1`] = `
{
  "params": [
    "["üë©üèΩ"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can check if null is in a JSON array: SQL Query for filter: {"settings.randomValues":{"$in":["apple",null]}} 1`] = `
{
  "params": [
    "null",
    "["apple"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_CONTAINS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues") IS NULL OR (JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?))",
}
`;

exports[`$in JSON Can check if null is in a JSON array: SQL Query for not match filter: {"settings.randomValues":{"$in":["apple"]}} 1`] = `
{
  "params": [
    "["apple"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search booleans in a JSON array: SQL Query for filter: {"settings.randomValues":{"$in":[false,true]}} 1`] = `
{
  "params": [
    "[false,true]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search booleans in a JSON array: SQL Query for filter: {"settings.randomValues":{"$in":[false]}} 1`] = `
{
  "params": [
    "[false]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search booleans in a JSON array: SQL Query for not match filter: {"settings.randomValues":{"$in":[]}} 1`] = `"always false"`;

exports[`$in JSON Can search booleans in a JSON array: SQL Query for not match filter: {"settings.randomValues":{"$in":[0]}} 1`] = `
{
  "params": [
    "[0]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search booleans in a JSON array: SQL Query for not match filter: {"settings.randomValues":{"$in":[true]}} 1`] = `
{
  "params": [
    "[true]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search numbers in a JSON array: SQL Query for filter: {"settings.randomValues":{"$in":[5,15]}} 1`] = `
{
  "params": [
    "[5,15]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search numbers in a JSON array: SQL Query for filter: {"settings.randomValues":{"$in":[5]}} 1`] = `
{
  "params": [
    "[5]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search numbers in a JSON array: SQL Query for filter: {"settings.randomValues":{"$in":[15]}} 1`] = `
{
  "params": [
    "[15]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search numbers in a JSON array: SQL Query for filter: {"settings.randomValues":{"$in":[99,15]}} 1`] = `
{
  "params": [
    "[99,15]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search numbers in a JSON array: SQL Query for not match filter: {"settings.randomValues":{"$in":[]}} 1`] = `"always false"`;

exports[`$in JSON Can search numbers in a JSON array: SQL Query for not match filter: {"settings.randomValues":{"$in":[6,10]}} 1`] = `
{
  "params": [
    "[6,10]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Can search numbers in a JSON array: SQL Query for not match filter: {"settings.randomValues":{"$in":[99,14]}} 1`] = `
{
  "params": [
    "[99,14]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON If a json array exists and does not contain null, it is not included when checking for null: SQL Query for filter: {"settings.randomValues":{"$in":[false]}} 1`] = `
{
  "params": [
    "[false]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON If a json array exists and does not contain null, it is not included when checking for null: SQL Query for not match filter: {"settings.randomValues":{"$in":["apple",null]}} 1`] = `
{
  "params": [
    "null",
    "["apple"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_CONTAINS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues") IS NULL OR (JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?))",
}
`;

exports[`$in JSON Including null also matches when a JSON path does not exist: SQL Query for filter: {"settings.randomValues":{"$in":["apple","banana","cherry",null]}} 1`] = `
{
  "params": [
    "null",
    "["apple","banana","cherry"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_CONTAINS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues") IS NULL OR (JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?))",
}
`;

exports[`$in JSON Including null also matches when a JSON path does not exist: SQL Query for not match filter: {"settings.randomValues":{"$in":["apple","banana","cherry"]}} 1`] = `
{
  "params": [
    "["apple","banana","cherry"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;

exports[`$in JSON Including null will also check on key existence for json arrays: SQL Query for filter: {"settings.randomValues":{"$in":["apple",null]}} 1`] = `
{
  "params": [
    "null",
    "["apple"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_CONTAINS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues") IS NULL OR (JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?))",
}
`;

exports[`$in JSON Including null will also check on key existence for json arrays: SQL Query for filter: {"settings.randomValues":{"$in":[null]}} 1`] = `
{
  "params": [
    "null",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_CONTAINS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues") IS NULL",
}
`;

exports[`$in JSON Including null will also check on key existence for json arrays: SQL Query for not match filter: {"settings.randomValues":{"$in":["apple"]}} 1`] = `
{
  "params": [
    "["apple"]",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_OVERLAPS(JSON_EXTRACT(\`test_table\`.\`settings\`,"$.randomValues"),?)",
}
`;
